### PR TITLE
Google SCC stats collection

### PR DESCRIPTION
**Pull request by paskal**
This PR adds Google Security Command center sources delay time gathering.

### CODE REVIEW

File: api/google.go [hunk 1]
```diff
@@ -63,3 +72,69 @@ func GetSCCHealthStatus(url string) int {
...
+		}
+	}
+	return result, nil
+}
+
+// GetSCCLatestEventTime return map of sources and their latest event update time difference with now
+// original: https://github.com/GoogleCloudPlatform/golang-samples/blob/master/securitycenter/findings/list_filtered_findings.go
+func GetSCCLatestEventTime(sources map[string]string) (map[string]time.Duration, error) {
+	result := make(map[string]time.Duration)
+	ctx := context.Background()
```
Comment (osdrv) [inline api/google.go]: Was any timeout meant here?

File: api/google.go [hunk 2]
```diff
@@ -63,3 +72,69 @@ func GetSCCHealthStatus(url string) int {
...
+	// process just one event with newest update date for every given source
+	for id, name := range sources {
+		req := &securitycenterpb.ListFindingsRequest{
+			Parent:   id,
+			OrderBy:  `event_time desc`,
+			PageSize: 1,
+		}
+		it := client.ListFindings(ctx, req)
+		findingsResult, err := it.Next()
+		if err != nil {
```
Comment (osdrv) [inline api/google.go]: I'd suggest handling iterator.Done and any other errors separately: not silently continuing

File: api/google.go [hunk 3]
```diff
@@ -63,3 +72,69 @@ func GetSCCHealthStatus(url string) int {
...
+	defer client.Close() // Closing the client safely cleans up background resources.
+
+	// process just one event with newest update date for every given source
+	for id, name := range sources {
+		req := &securitycenterpb.ListFindingsRequest{
+			Parent:   id,
+			OrderBy:  `event_time desc`,
+			PageSize: 1,
+		}
+		it := client.ListFindings(ctx, req)
```
Comment (osdrv) [inline api/google.go:135]: The semantics of this call tells there might be more than 1 result, which doesn't make sense from the business logic perspective: it is getting an iterator where contextually might be 0, 1, 2 and more elements. A more explicit single-unit container is more preferable here.

File: api/google_test.go [hunk 4]
```diff
@@ -77,5 +77,19 @@ func TestGetGoogleSCCHealthStatus(t *testing.T) {
 		status := GetSCCHealthStatus(x.serverURL)
 		assert.Equal(t, x.status, status, "Test case %d status check failed", i)
 	}
+}
+
+func TestGetSCCLatestEventTime(t *testing.T) {
```
Comment (osdrv) [inline api/google_test.go]: Not sure the test name reflects what's happening inside

File: api/google_test.go [hunk 5]
```diff
@@ -77,5 +77,19 @@ func TestGetGoogleSCCHealthStatus(t *testing.T) {
...
+	x, err := GetSCCLatestEventTime(nil)
+	assert.Nil(t, x)
+	assert.Error(t, err, "no authentication present should result in error")
+}
 
+func TestGetSCCSourcesByName(t *testing.T) {
+	x, err := GetSCCSourcesByName("", "bad_regex(")
```
Comment (osdrv) [inline api/google_test.go:92]: This test also exercises only 1 specific case and it's name doesn't reflect that. Consider defining a range of test cases covering corner cases in a table structure and testing them sequentially within a single test func.

**Review by osdrv — CHANGES_REQUESTED** [review]

**Comment by grzkv** [comment]
I can see that this can be merged without approval. I'd suggest changing the repo policy to require at least one approval for PRs + protect the `master` branch. This will guarantee quality.

File: api/google.go (hunk 1 above)
Comment (paskal) [reply api/google.go]: Not originally, but it makes sense to do: I've added 20s timeout to both.

File: api/google.go (hunk 2 above)
Comment (paskal) [reply api/google.go]: There might be no events to fetch and then no information will be available. The current behavior is desired, maybe I should reflect this in function documentation.

File: README.md [hunk 6]
```diff
@@ -19,9 +19,11 @@ docker-compose run metrics --help
 | prisma_api_url          | PRISMA_API_URL          | https://api.eu.prismacloud.io | Prisma API key                   |
 | prisma_api_key          | PRISMA_API_KEY          |                          | Prisma API key                        |
 | prisma_api_password     | PRISMA_API_PASSWORD     |                          | Prisma API password                   |
+| scc_org_id              | SCC_ORG_ID              |                          | Google SCC numeric organisation ID    |
+| scc_sources_regex       | SCC_SOURCES_REGEX       | `.`                      | Google SCC sources Display Name filter regexp |
 | graphite_host           | GRAPHITE_HOST           |                          | Graphite hostname                     |
 | graphite_port           | GRAPHITE_PORT           | `2003`                   | Graphite port                         |
-| graphite_prefix         | GRAPHITE_PREFIX         |                          | Graphite port                         |
+| graphite_prefix         | GRAPHITE_PREFIX         |                          | Graphite metrics prefix               |
```
Comment (osdrv) [inline README.md]: As a side observer, I'm a bit confused what's the difference between `graphite_prefix` and `compliance_prefix`. Not sure if it's expected that a user of this code should come with a pre-knowledge of this specific difference: if it's not the case, it's worth elaborating in the comment at the left side.

File: README.md [hunk 7]
```diff
@@ -34,6 +36,9 @@ Collected metrics list:
   - API health status ([SLA](https://www.paloaltonetworks.com/resources/datasheets/prisma-public-cloud-service-level-agreement))
 - [Google Security Command Center](https://cloud.google.com/security-command-center/):
   - [health status](https://status.cloud.google.com/)
+  - specified sources newest event update time (for monitoring [Forseti](https://forsetisecurity.org/) alerting delay).
```
Comment (osdrv) [inline README.md]: Consider rephrasing this sentence

File: api/google.go [hunk 8]
```diff
@@ -1,4 +1,5 @@
 // Copyright 2019 Booking.com
+// Copyright 2019 Google LLC
```
Comment (osdrv) [inline api/google.go]: That's concerning. I'm afraid this change needs an approval of Booking.com legals.

File: api/google.go [hunk 9]
```diff
@@ -63,3 +75,75 @@ func GetSCCHealthStatus(url string) int {
...
+// GetSCCSourcesByName returns Security Command Center sources for given numeric orgID,
+// filtered by name by given regex
+// original: https://github.com/GoogleCloudPlatform/golang-samples/blob/master/securitycenter/findings/list_sources.go
+func GetSCCSourcesByName(orgID string, nameRegex string) (map[string]string, error) {
+	regex, err := regexp.Compile(nameRegex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error compiling nameRegex")
+	}
+	// Instantiate a context and a security service client to make API calls.
+	ctx, cancel := context.WithTimeout(context.Background(), apiTimeout)
```
Comment (osdrv) [inline api/google.go:87]: Good job! I appreciate you're following up on the previous comment!

File: api/google.go [hunk 10]
```diff
@@ -63,3 +75,75 @@ func GetSCCHealthStatus(url string) int {
...
+	defer client.Close() // Closing the client safely cleans up background resources.
+
+	req := &securitycenterpb.ListSourcesRequest{
+		Parent: fmt.Sprintf("organizations/%s", orgID),
+	}
+	it := client.ListSources(ctx, req)
+	result := map[string]string{}
+	for {
+		source, err := it.Next()
+		if err == iterator.Done {
```
Comment (osdrv) [inline api/google.go:102]: Nice looking now, the separation of errors totally make sense!

**Review by osdrv — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (grzkv) [reply README.md]: I agree with @icanhazbroccoli Also, `compliance_prefix` feels like something very specific to a single use-case.

File: api/google.go (hunk 8 above)
Comment (grzkv) [reply api/google.go]: +1 @icanhazbroccoli

File: api/google.go [hunk 11]
```diff
@@ -63,3 +72,69 @@ func GetSCCHealthStatus(url string) int {
...
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "it.Next")
```
Comment (grzkv) [inline api/google.go]: I would enjoy a bit more background for this error. It will be a bit unclear when met by someone who did not write the code.

**Review by grzkv — CHANGES_REQUESTED** [review]

**Review by kostty — APPROVED** [review]
Looks good to me

**Review by osdrv — APPROVED** [review]
